core-tests/shared/src/test/scala/zio/stm/TArraySpec.scala

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package zio.stm
+
+import zio.{ FiberFailure, UIO, ZIO, ZIOBaseSpec }
+import zio.test._
+import zio.test.Assertion._
+import TArraySpecUtil._
+
+object TArraySpec
+    extends ZIOBaseSpec(
+      suite("TArraySpec")(
+        suite("apply")(
+          testM("happy-path") {
+            for {
+              tArray <- makeTArray(1)(42)
+              value  <- tArray(0).commit
+            } yield assert(value, equalTo(42))
+          },
+          testM("dies with ArrayIndexOutOfBounds when index is out of bounds $applyOutOfBounds") {
+            assertM(for {
+              tArray <- makeTArray(1)(42)
+              _      <- tArray(-1).commit
+            } yield (), throws(equalTo(FiberFailure)))
```

Comment:
This doesn't work. Not sure how to write that assert.

Reply:
Expected behavior is that `Fiber` running `TArray#appy(INVALID_INDEX)` will DIE, because it's considered an fatal error. So I _guess_ there are two things to change:
* use `.run` combinator to get some result or exit value from apply
* because we want to ensure DIE of a fiber, then `dies` instead of `throws`.

Reply:
It's quite more complicated, since `new ArrayIndexOutOfBoundsException(-1)` is not equal to `new ArrayIndexOutOfBoundsException(-1)`, we need some other assertion. I didn't manage to use `isSubtype` - I don't know how :man_shrugging: 

I managed to fix this with creating custom assertion:
`Assertion.assertion[Throwable]("isArrayIndexOutOfBoundsException")()(_.isInstanceOf[ArrayIndexOutOfBoundsException]))`

A note to my first comment: `run` needs to be applied on a ZIO that contains call to apply(-1) inside. Because calling `tArray(-1)` is analogous to `Future.succeed(throw new Excepion)`.

Reply:
Here is what I have so far:
```
val result = for {
    tArray <- makeTArray(1)(42)
    r      <- ZIO.effect(tArray(-1)).run
} yield r.toEither
assertM(result, isLeft(Assertion.assertion[Throwable]("isArrayIndexOutOfBoundsException")()(_.isInstanceOf[ArrayIndexOutOfBoundsException])))
```
I get the following error while running the test:
```
[info]     - dies with ArrayIndexOutOfBounds when index is out of bounds
[info]       Left(zio.FiberFailure: Fiber failed.
[info] A checked error was not handled.
[info] java.lang.ArrayIndexOutOfBoundsException: Array index out of range: -1
[info] 	at zio.stm.TArray$.apply$extension(TArray.scala:27)
[info] 	at zio.stm.TArraySpec$$anonfun$$lessinit$greater$1.$anonfun$new$4(TArraySpec.scala:36)
[info] 	at zio.stm.TArraySpec$$anonfun$$lessinit$greater$1.$anonfun$new$4$adapted(TArraySpec.scala:36)
[info] 	at zio.internal.FiberContext.evaluateNow(FiberContext.scala:403)
[info] 	at zio.internal.FiberContext.$anonfun$fork$1(FiberContext.scala:595)
[info] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[info] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[info] 	at java.lang.Thread.run(Thread.java:745)
```
I couldn't get type inference to work with `dies` that's why I used `toEither`.

Reply:
I made it work with the help @adamgfraser and your suggestion with the custom assertion. Thank you!